rfc/0077-zone-aware-replica-reads.md

```diff
@@ -0,0 +1,183 @@
+`ReadPreference::NO_PREFERENCE`.
+
+```
+class GetAllReplicasOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+```
+class GetAnyReplicaOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+```
+class LookupInAllReplicasOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+```
+class LookupInAnyReplicaOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+In all failure cases, when the SDK cannot handle operation, it must return `102
+DocumentUnretrievable` with human-readable explanation whenever it is possible
+that explains the details.
+
+# Open Questions
+
+## Q1
+
+What the SDK should do if the local group does not have any replica configured
+for the key, but the User request `Local-Only` strategy? Consider the following
+cases:
+
+- number of replicas is 1, number of zones is 3. Neither active, nor replica is
+  in the local group.
+- number of replicas is 1, number of zones is 3. Only active node is in the
+  local group.
+- number of replicas is 1, number of zones is 3. Only replica node is in the
+  local group.
+
+Shall we warn users when locality restriction turns request into just GET,
+because only active node is in local group?
+
+## Q2
+
+Do we need to implement management API to create/retrieve server groups and
+change node belonging to particular group?
+
+## Q3
+
+All currently supported server versions allow to configure server groups, but
+only 7.6.1 announce this information in configuration.
+[MB-60835](https://issues.couchbase.com/browse/MB-60835) How the SDK should
+behave with older server? Do we need to fail fast or silently fall back to
+current behavior?
+
+## Q4
+
+Is there explicit statement from the server team that mixed-version cluster will
+not announce feature until all nodes migrated to 7.6.1?
+
+## Q5
+
+Moving nodes between groups generates configuration, but does not really moves
+data. The user might forget to trigger rebalance, and the SDK will try to do
+Local-Only or Local-First strategy, while the operations are still expensive.
+
+## Q6
+
+How the SDK should behave during rebalance? It cannot guarantee cheap traffic
+based on availability. Should we retry operations if the configuration says that
+replica has moved to different group?
+
+## Q7
+
+In case of more than two server groups configured, do we need to distinguish
+between non-local groups? Right now there is no weights associated with the
+groups, so we don't have enough information to reason about cost of
+communication with non-local groups.
+
+## Q8
+
+Do we want to update GetResult and LookupInResult with the property, which would
+tell the user the name of the availability zone, where serving node has been
+deployed upon generating response?
```

Comment:
This could be a neat solution to a lot of the concerns above.  If we tell the user where we got the replica from, they could add their own warning if it doesn't match the server group they expected, rather than us needing to spam.

Reply:
Lets keep first iteration simple, and add it later if we find real use-case or demand